lib/rom/sql/attribute.rb

```diff
@@ -14,6 +14,49 @@ class Attribute < ROM::Schema::Attribute
       # Error raised when an attribute cannot be qualified
       QualifyError = Class.new(StandardError)
 
+      # Type-specific methods
+      #
+      # @api public
+      module TypeExtensions
+        class << self
+          # Gets extensions for a type
+          #
+          # @param [Dry::Types::Type] type
+          #
+          # @return [Hash]
+          #
+          # @api public
+          def [](type)
+            unwrapped = type.optional? ? type.right : type
```

Comment:
yeah but...technically speaking somebody could do `attribute :foo, Types::String.optional | Types::Int.optional` for whatever reason (like using sqlite 😆). Just thinking that maybe adding a simple check and raising an arg error if `type.right` is also optional would reduce the risk of having awkward backtraces.

Anyhow, just thinking out loud, let's leave it and wait how it works out.